test/e2e/secrets_test.go

```diff
@@ -74,59 +74,79 @@ func testUserData(t *testing.T) {
+	// until the Machine is back up.
+	log.Printf("waiting for Machine nodes to be recreated after userdata update")
+	assert.NoError(t, tc.waitForNewMachineNodes(), "error waiting for Machine nodes to be reconfigured")
+	assert.NoError(t, tc.waitForValidUserData(validUserDataSecret), "error waiting for valid userdata")
+}
+
+// waitForNewMachineNodes is a returns an error if waitForWindowsNodes returns the same Machine backed nodes
+func (tc *testContext) waitForNewMachineNodes() error {
+	var oldNodes []string
+	for _, node := range gc.machineNodes {
+		oldNodes = append(oldNodes, node.GetName())
+	}
 
-			// wait for userData secret creation / update to take effect.
-			err := wait.Poll(retry.Interval, retry.ResourceChangeTimeout, func() (done bool, err error) {
-				secretInstance, err = tc.client.K8s.CoreV1().Secrets(clusterinfo.MachineAPINamespace).Get(context.TODO(), "windows-user-data", meta.GetOptions{})
-				if err != nil {
-					if apierrors.IsNotFound(err) {
-						log.Printf("still waiting for user data secret: %v", err)
-						return false, nil
-					}
-					log.Printf("error listing secrets: %v", err)
-					return false, nil
-				}
-				if string(validUserDataSecret.Data["userData"][:]) != string(secretInstance.Data["userData"][:]) {
-					return false, nil
-				}
-				return true, nil
-			})
-			require.NoError(t, err, "could not find a valid userData secret in the namespace : %v", secretInstance.Namespace)
-		})
+	// waitForWindowsNodes will re-populate gc.machineNodes with the configured nodes found
+	err := tc.waitForWindowsNodes(gc.numberOfMachineNodes, false, false, false)
+	if err != nil {
+		return err
+	}
+	for _, newNode := range gc.machineNodes {
+		for _, oldNode := range oldNodes {
+			if newNode.GetName() == oldNode {
+				return errors.Wrapf(err, "node %s is not a new Node", oldNode)
+			}
+		}
 	}
+	return nil
+}
+
+// testUserDataRegeneration tests that the userdata will be created by WMCO if deleted by a user
+func (tc *testContext) testUserDataRegeneration(t *testing.T) {
+	validUserDataSecret, err := tc.client.K8s.CoreV1().Secrets(clusterinfo.MachineAPINamespace).Get(context.TODO(),
+		secrets.UserDataSecret, meta.GetOptions{})
+	require.NoError(t, err, "could not find Windows userData secret in required namespace")
+	err = tc.client.K8s.CoreV1().Secrets(clusterinfo.MachineAPINamespace).Delete(context.TODO(),
+		secrets.UserDataSecret, meta.DeleteOptions{})
+	require.NoError(t, err, "could not delete userData secret")
+	assert.NoError(t, tc.waitForValidUserData(validUserDataSecret))
+}
+
+// waitForValidUserData waits until the userdata secret exists with expected content
+func (tc *testContext) waitForValidUserData(expected *core.Secret) error {
+	return wait.Poll(retry.Interval, retry.ResourceChangeTimeout, func() (done bool, err error) {
+		actualData, err := tc.getUserDataContents()
+		if err != nil {
+			if apierrors.IsNotFound(err) {
+				log.Printf("still waiting for user data secret: %v", err)
```

Comment:
is this a debug log or should be in the final version of code?

Reply:
It should be there, it isn't a new log